File: Longest Substring Without Repeating Characters - Leetcode 3/Longest Substring Without Repeating Characters - Leetcode 3.js
// brute force
// Time Complexity: O(n^3)
// Space Complexity: O(n)
var lengthOfLongestSubstring = function(s) {

    let longest = 0;
    let n = s.length;
    for (let i = 0; i < n; ++i) {
        for (let substr_len = 1; i + substr_len <= n; ++substr_len) {
            var seen = new Set();
            for (let j = i; j < i + substr_len; ++j) {
                seen.add(s[j]);
            }
            if (seen.size == substr_len) {
                longest = Math.max(longest, substr_len);
            }
        }
    }

    return longest;
};


var lengthOfLongestSubstring = function(s) {
    const set = new Set();
    let l = 0, longest = 0;

    for (let r = 0; r < s.length; r++) {
        while (set.has(s[r])) {
            set.delete(s[l]);
            l++;
        }

        longest = Math.max(longest, r - l + 1);
        set.add(s[r]);
    }

    return longest;
};

File: Longest Substring Without Repeating Characters - Leetcode 3/Longest Substring Without Repeating Characters - Leetcode 3.java
import java.util.HashSet;


// brute force
// Time Complexity: O(n^3)
// Space Complexity: O(n)
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        int longest = 0;
        int n = s.length();
        for (int i = 0; i < n; i++) {
            for (int substr_len = 1; i + substr_len <= n; ++substr_len) {
                HashSet<Character> seen = new HashSet<>();
                for (int j = i; j < i + substr_len; ++j) 
                    seen.add(s.charAt(j));

                if (seen.size() == substr_len)
                    longest = Math.max(longest, substr_len);
            }
        }

        return longest;
    }
}




public class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int l = 0, longest = 0;

        for (int r = 0; r < s.length(); r++) {
            while (set.contains(s.charAt(r))) {
                set.remove(s.charAt(l));
                l++;
            }

            longest = Math.max(longest, r - l + 1);
            set.add(s.charAt(r));
        }

        return longest;
    }
}
